In [1]:
import re

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 11: Cosmic Expansion ---</h2><p>You continue following signs for "Hot Springs" and eventually come across an <a href="https://en.wikipedia.org/wiki/Observatory" target="_blank">observatory</a>. The Elf within turns out to be a researcher studying cosmic expansion using the giant telescope here.</p>
<p>He doesn't know anything about the missing machine parts; he's only visiting for this research project. However, he confirms that the hot springs are the next-closest area likely to have people; he'll even take you straight there once he's done with today's observation analysis.</p>
<p>Maybe you can help him with the analysis to speed things up?</p>
<p>The researcher has collected a bunch of data and compiled the data into a single giant <em>image</em> (your puzzle input). The image includes <em>empty space</em> (<code>.</code>) and <em>galaxies</em> (<code>#</code>). For example:</p>
<pre><code>...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....
</code></pre>
<p>The researcher is trying to figure out the sum of the lengths of the <em>shortest path between every pair of galaxies</em>. However, there's a catch: the universe expanded in the time it took the light from those galaxies to reach the observatory.</p>
<p>Due to something involving gravitational effects, <em>only some space expands</em>. In fact, the result is that <em>any rows or columns that contain no galaxies</em> should all actually be twice as big.</p>
<p>In the above example, three columns and two rows contain no galaxies:</p>
<pre><code>   v  v  v
 ...#......
 .......#..
 #.........
&gt;..........&lt;
 ......#...
 .#........
 .........#
&gt;..........&lt;
 .......#..
 #...#.....
   ^  ^  ^
</code></pre>
<p>These rows and columns need to be <em>twice as big</em>; the result of cosmic expansion therefore looks like this:</p>
<pre><code>....#........
.........#...
#............
.............
.............
........#....
.#...........
............#
.............
.............
.........#...
#....#.......
</code></pre>
<p>Equipped with this expanded universe, the shortest path between every pair of galaxies can be found. It can help to assign every galaxy a unique number:</p>
<pre><code>....1........
.........2...
3............
.............
.............
........4....
.5...........
............6
.............
.............
.........7...
8....9.......
</code></pre>
<p>In these 9 galaxies, there are <em>36 pairs</em>. Only count each pair once; order within the pair doesn't matter. For each pair, find any shortest path between the two galaxies using only steps that move up, down, left, or right exactly one <code>.</code> or <code>#</code> at a time. (The shortest path between two galaxies is allowed to pass through another galaxy.)</p>
<p>For example, here is one of the shortest paths between galaxies <code>5</code> and <code>9</code>:</p>
<pre><code>....1........
.........2...
3............
.............
.............
........4....
.5...........
.##.........6
..##.........
...##........
....##...7...
8....9.......
</code></pre>
<p>This path has length <code><em>9</em></code> because it takes a minimum of <em>nine steps</em> to get from galaxy <code>5</code> to galaxy <code>9</code> (the eight locations marked <code>#</code> plus the step onto galaxy <code>9</code> itself). Here are some other example shortest path lengths:</p>
<ul>
<li>Between galaxy <code>1</code> and galaxy <code>7</code>: 15</li>
<li>Between galaxy <code>3</code> and galaxy <code>6</code>: 17</li>
<li>Between galaxy <code>8</code> and galaxy <code>9</code>: 5</li>
</ul>
<p>In this example, after expanding the universe, the sum of the shortest path between all 36 pairs of galaxies is <code><em>374</em></code>.</p>
<p>Expand the universe, then find the length of the shortest path between every pair of galaxies. <em>What is the sum of these lengths?</em></p>
</article>


In [2]:
image = """
...#......
.......#..
#.........
..........
......#...
.#........
.........#
..........
.......#..
#...#.....
"""

expected_image = """
....#........
.........#...
#............
.............
.............
........#....
.#...........
............#
.............
.............
.........#...
#....#.......

"""

In [3]:
from itertools import product, zip_longest
from tabulate import tabulate


def sum_min_distances(image: str) -> int:
    graph = create_universe(image)
    distance_sum = 0
    points = get_points(graph)
    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            distance_sum += mahatten_distance(points[i], points[j])
    return distance_sum


def create_universe(image: str) -> list[list[int]]:
    graph = [tuple(l) for l in image.strip().splitlines()]
    graph = transpose(graph)
    graph = expand(graph)
    graph = transpose(graph)
    return expand(graph)


def transpose(graph: list[list[str]]) -> list[list[str]]:
    return list(zip(*graph))


def expand(graph: list[list[str]]) -> list[list[str]]:
    expanded = []
    for line in graph:
        expanded.append(line)
        if all(cell == "." for cell in line):
            expanded.append(line)
    return expanded


def get_points(graph: list[list[str]]) -> list[tuple[int, int]]:
    return [
        (r, c)
        for r, c in product(range(len(graph)), range(len(graph[0])))
        if graph[r][c] == "#"
    ]


def mahatten_distance(p1: tuple[int, int], p2: tuple[int, int]) -> int:
    return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])


# sum_min_distances(image)


expected = [tuple(l) for l in expected_image.strip().splitlines()]
actual = create_universe(image)
assert actual == expected

assert sum_min_distances(image) == 374

In [4]:
with open("../input/day11.txt") as f:
    print(sum_min_distances(f.read()))

9599070


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>9599070</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The galaxies are much <em>older</em> (and thus much <em>farther apart</em>) than the researcher initially estimated.</p>
<p>Now, instead of the expansion you did before, make each empty row or column <em><span title="And you have to have your pinky near your mouth when you do it.">one million</span> times</em> larger. That is, each empty row should be replaced with <code>1000000</code> empty rows, and each empty column should be replaced with <code>1000000</code> empty columns.</p>
<p>(In the example above, if each empty row or column were merely <code>10</code> times larger, the sum of the shortest paths between every pair of galaxies would be <code><em>1030</em></code>. If each empty row or column were merely <code>100</code> times larger, the sum of the shortest paths between every pair of galaxies would be <code><em>8410</em></code>. However, your universe will need to expand far beyond these values.)</p>
<p>Starting with the same initial image, expand the universe according to these new rules, then find the length of the shortest path between every pair of galaxies. <em>What is the sum of these lengths?</em></p>
</article>

<p>Although it hasn't changed, you can still <a href="11/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+Part+One+of+%22Cosmic+Expansion%22+%2D+Day+11+%2D+Advent+of+Code+2023&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F11&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+Part+One+of+%22Cosmic+Expansion%22+%2D+Day+11+%2D+Advent+of+Code+2023+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F11';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>


In [5]:
from itertools import product, zip_longest
from tabulate import tabulate


def sum_min_distances(image: str, factor=2) -> int:
    graph, expanded_rows, expanded_cols = create_universe(image, factor)
    points = get_points(graph, expanded_rows, expanded_cols)
    distance_sum = 0
    for i in range(len(points)):
        for j in range(i + 1, len(points)):
            distance_sum += mahatten_distance(points[i], points[j])
    return distance_sum


def transpose(graph: list[list[str]]) -> list[list[str]]:
    return list(zip(*graph))


def create_universe(image: str, factor) -> list[list[int]]:
    graph = [tuple(l) for l in image.strip().splitlines()]
    expanded_rows = [0] * len(graph)
    expanded_rows[0] = factor if all(cell == "." for cell in graph[0]) else 1

    for i in range(1, len(graph)):
        if all(cell == "." for cell in graph[i]):
            expanded_rows[i] = expanded_rows[i - 1] + factor
        else:
            expanded_rows[i] = expanded_rows[i - 1] + 1

    graph_T = transpose(graph)
    expanded_cols = [1] * len(graph_T)
    expanded_cols[0] = factor if all(cell == "." for cell in graph_T[0]) else 1
    for i in range(1, len(graph_T)):
        if all(cell == "." for cell in graph_T[i]):
            expanded_cols[i] = expanded_cols[i - 1] + factor
        else:
            expanded_cols[i] = expanded_cols[i - 1] + 1

    return graph, expanded_rows, expanded_cols


def get_points(
    graph: list[list[str]], expanded_rows: list[int], expanded_cols: list[int]
) -> list[tuple[int, int]]:
    return [
        (expanded_rows[r], expanded_cols[c])
        for r, c in product(range(len(graph)), range(len(graph[0])))
        if graph[r][c] == "#"
    ]


def mahatten_distance(p1: tuple[int, int], p2: tuple[int, int]) -> int:
    return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])


assert sum_min_distances(image, 2) == 374
assert sum_min_distances(image, 10) == 1030
assert sum_min_distances(image, 100) == 8410

In [6]:
with open("../input/day11.txt") as f:
    print(sum_min_distances(f.read(), 1_000_000))

842645913794


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>842645913794</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2023">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="11/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Cosmic+Expansion%22+%2D+Day+11+%2D+Advent+of+Code+2023&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F11&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Cosmic+Expansion%22+%2D+Day+11+%2D+Advent+of+Code+2023+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2023%2Fday%2F11';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
